File: permissions.py
from __future__ import annotations

import fnmatch
from dataclasses import dataclass, field
from enum import Enum
from pathlib import Path


class PermissionMode(Enum):
    DEFAULT = "default"  # prompt before writes + shell
    AUTO = "auto"        # allow everything (CI / sandboxed)
    PLAN = "plan"        # block all writes — read-only


class PermissionDenied(Exception):
    pass


class PermissionRequired(Exception):
    """Raised in DEFAULT mode — caller handles the prompt."""
    def __init__(self, tool: str, args: dict):
        self.tool = tool
        self.args = args
        super().__init__(f"Permission required for {tool}")


@dataclass
class PathRule:
    pattern: str
    allow: bool


_SENSITIVE_PATHS = [
    "~/.ssh/*",
    "~/.aws/*",
    "~/.config/*/credentials",
    "~/.gnupg/*",
    "~/.netrc",
]

_WRITE_TOOLS = {"write_file", "edit_file"}
_EXEC_TOOLS = {"run_bash"}


@dataclass
class PermissionPolicy:
    mode: PermissionMode = PermissionMode.DEFAULT
    path_rules: list[PathRule] = field(default_factory=list)
    denied_commands: list[str] = field(default_factory=list)

    @classmethod
    def from_mode(cls, mode: str) -> "PermissionPolicy":
        return cls(mode=PermissionMode(mode))

    def check(self, tool: str, args: dict) -> None:
        """Raise PermissionDenied or PermissionRequired as appropriate."""
        if self.mode == PermissionMode.PLAN and tool in _WRITE_TOOLS | _EXEC_TOOLS:
            raise PermissionDenied(f"{tool} blocked in PLAN mode")

        # Sensitive path guard — applies in all modes
        path = args.get("path") or args.get("file_path") or ""
        if path:
            self._check_path(path)

        # Check custom path rules
        for rule in self.path_rules:
            expanded = str(Path(rule.pattern).expanduser())
            if fnmatch.fnmatch(path, expanded):
                if not rule.allow:
                    raise PermissionDenied(f"Path blocked by rule: {rule.pattern}")
                break  # explicit allow stops further checks

        # Denied commands prefix check
        if tool == "run_bash":
            cmd = args.get("command", "")
            for denied in self.denied_commands:
                if cmd.startswith(denied):
                    raise PermissionDenied(f"Command blocked: starts with '{denied}'")

        # DEFAULT mode: write/shell requires caller approval
        if self.mode == PermissionMode.DEFAULT:
            if tool in _WRITE_TOOLS | _EXEC_TOOLS:
                raise PermissionRequired(tool, args)

    def _check_path(self, path: str) -> None:
        for pattern in _SENSITIVE_PATHS:
            expanded = str(Path(pattern).expanduser())
            if fnmatch.fnmatch(path, expanded):
                raise PermissionDenied(f"Access to sensitive path blocked: {path}")

File: src/permissions.ts
import micromatch from "micromatch";
import { homedir } from "os";
import { join } from "path";

const HOME = homedir();

export enum PermissionMode {
  DEFAULT = "default",
  AUTO = "auto",
  PLAN = "plan",
}

export class PermissionDenied extends Error {
  constructor(message: string) {
    super(message);
    this.name = "PermissionDenied";
  }
}

export class PermissionRequired extends Error {
  constructor(
    public readonly tool: string,
    public readonly args: Record<string, unknown>,
  ) {
    super(`Permission required for ${tool}`);
    this.name = "PermissionRequired";
  }
}

export interface PathRule {
  pattern: string;
  allow: boolean;
}

const SENSITIVE_PATHS: string[] = [
  join(HOME, ".ssh", "*"),
  join(HOME, ".aws", "*"),
  join(HOME, ".config", "*", "credentials"),
  join(HOME, ".gnupg", "*"),
  join(HOME, ".netrc"),
];

const WRITE_TOOLS = new Set(["write_file", "edit_file"]);
const EXEC_TOOLS = new Set(["run_bash"]);

export class PermissionPolicy {
  constructor(
    public readonly mode: PermissionMode = PermissionMode.DEFAULT,
    public readonly pathRules: PathRule[] = [],
    public readonly deniedCommands: string[] = [],
  ) {}

  static fromMode(mode: string): PermissionPolicy {
    return new PermissionPolicy(mode as PermissionMode);
  }

  check(tool: string, args: Record<string, unknown>): void {
    if (
      this.mode === PermissionMode.PLAN &&
      (WRITE_TOOLS.has(tool) || EXEC_TOOLS.has(tool))
    ) {
      throw new PermissionDenied(`${tool} blocked in PLAN mode`);
    }

    const path = String(args["path"] ?? args["file_path"] ?? "");
    if (path) {
      this._checkPath(path);
    }

    // custom path rules
    for (const rule of this.pathRules) {
      const expanded = expandUser(rule.pattern);
      if (micromatch.isMatch(path, expanded)) {
        if (!rule.allow) {
          throw new PermissionDenied(`Path blocked by rule: ${rule.pattern}`);
        }
        break;
      }
    }

    // denied commands prefix check
    if (tool === "run_bash") {
      const cmd = String(args["command"] ?? "");
      for (const denied of this.deniedCommands) {
        if (cmd.startsWith(denied)) {
          throw new PermissionDenied(`Command blocked: starts with '${denied}'`);
        }
      }
    }

    // DEFAULT mode: write/shell requires caller approval
    if (this.mode === PermissionMode.DEFAULT) {
      if (WRITE_TOOLS.has(tool) || EXEC_TOOLS.has(tool)) {
        throw new PermissionRequired(tool, args);
      }
    }
  }

  private _checkPath(path: string): void {
    for (const pattern of SENSITIVE_PATHS) {
      if (micromatch.isMatch(path, pattern)) {
        throw new PermissionDenied(`Access to sensitive path blocked: ${path}`);
      }
    }
  }
}

function expandUser(p: string): string {
  if (p.startsWith("~/")) return join(HOME, p.slice(2));
  if (p === "~") return HOME;
  return p;
}
